### DISCUSSION TITLE
v4.12.9 - Added RestRequest and RestResponse Logging

### DISCUSSION THREAD

**Discussion by jongpie** [jongpie/NebulaLogger#622] [Announcements]
Thanks to @diegocomi for suggesting this enhancement!

# Core Unlocked Package Changes

## Added logging support for instances of `RestRequest` and `RestResponse`

- Resolved #514 (suggested by @diegocomi) in PR #613 by introducing new features & fields to support logging instances of [`System.RestRequest`](https://developer.salesforce.com/docs/atlas.en-us.apexref.meta/apexref/apex_methods_system_restrequest.htm) and [`System.RestResponse`](https://developer.salesforce.com/docs/atlas.en-us.apexref.meta/apexref/apex_methods_system_restresponse.htm) that are used for [Apex REST services](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_rest_intro.htm):

  - Added 2 new `global` instance methods in `LogEntryBuilder` to provide a way to log instances of `System.RestRequest` and `System.RestResponse`:
    1. `global LogEntryEventBuilder setRestRequestDetails(System.RestRequest request);`
    2. `global LogEntryEventBuilder setRestResponseDetails(System.RestResponse response);`

  - Added 2 new `LoggerParameter__mdt` records to control logging of request & response headers when using the new  builder methods. These `LoggerParameter__mdt` records are equivalent to the record `StoreHttpResponseHeaderValues` (introduced in [Nebula Logger `v4.11.8`](https://github.com/jongpie/NebulaLogger/releases/tag/v4.11.8)) for `System.HttpResponse` logging:
    1. `StoreRestRequestHeaderValues` - this metadata records controls if the request's header values are stored in new fields `LogEntryEvent__e.RestRequestHeaders__c` and `LogEntryEvent__e.RestrequestHeaders__c`. By default, it's enabled.
    2. `StoreRestResponseHeaderValues` - this metadata records controls if the response's header values are stored in new fields `LogEntryEvent__e.RestResponseHeaders__c` and `LogEntryEvent__e.RestResponseHeaders__c`. By default, it's enabled.

  - Added 9 new fields for `System.RestRequest` data on `LogEntryEvent__e` platform event
    1. `LogEntryEvent__e.RestRequestBodyMasked__c`
    4. `LogEntryEvent__e.RestRequestBody__c`
    5. `LogEntryEvent__e.RestRequestHeaderKeys__c`
    6. `LogEntryEvent__e.RestRequestHeaders__c`
    7. `LogEntryEvent__e.RestRequestMethod__c`
    8. `LogEntryEvent__e.RestRequestParameters__c`
    9. `LogEntryEvent__e.RestRequestRemoteAddress__c`
    10. `LogEntryEvent__e.RestRequestResourcePath__c`
    11. `LogEntryEvent__e.RestRequestUri__c`

  - Added 5 new fields for `System.RestResponse` data on `LogEntryEvent__e` platform event
    1. `LogEntryEvent__e.RestResponseBodyMasked__c`
    2. `LogEntryEvent__e.RestResponseBody__c`
    5. `LogEntryEvent__e.RestResponseHeaderKeys__c`
    6. `LogEntryEvent__e.RestResponseHeaders__c`
    7. `LogEntryEvent__e.RestResponseStatusCode__c`

  - Added 12 new fields for `System.RestResponse` data on `LogEntry__c` custom object
    1. `LogEntry__c.HasRestRequestBody__c`
    2. `LogEntry__c.HasRestRequestHeaderKeys__c`
    3. `LogEntry__c.HasRestRequestHeaders__c`
    4. `LogEntry__c.RestRequestBody__c`
    5. `LogEntry__c.RestRequestBodyMasked__c`
    6. `LogEntry__c.RestRequestHeaderKeys__c`
    7. `LogEntry__c.RestRequestHeaders__c`
    8. `LogEntry__c.RestRequestMethod__c`
    9. `LogEntry__c.RestRequestParameters__c`
    10. `LogEntry__c.RestRequestRemoteAddress__c`
    11. `LogEntry__c.RestRequestResourcePath__c`
    12. `LogEntry__c.RestRequestUri__c`

  - Added 8 new fields for `System.RestResponse` data on `LogEntry__c` custom object
    1. `LogEntry__c.HasRestResponseBody__c`
    2. `LogEntry__c.HasRestResponseHeaderKeys__c`
    4. `LogEntry__c.HasRestResponseHeaders__c`
    5. `LogEntry__c.RestResponseBodyMasked__c`
    6. `LogEntry__c.RestResponseBody__c`
    8. `LogEntry__c.RestResponseHeaderKeys__c`
    9. `LogEntry__c.RestResponseHeaders__c`
    10. `LogEntry__c.RestResponseStatusCode__c`

  - Updated permission sets `LoggerAdmin`, `LoggerLogViewer`, and `LoggerEndUser` to add access to the new custom fields on `LogEntry__c`
  - Add 2 new sections on `LogEntryRecordPage` flexipage (shown in the screenshot below)
    - `Apex REST Service Request` section shows the corresponding fields for `RestRequest` data (when populated)
    - `Apex REST Service Response` section shows the corresponding fields for `RestResponse` data (when populated)
  - Renamed 2 existing sections on `LogEntryRecordPage` flexipage to help avoid confusion with the 2 new sections `Apex Rest Service Request` and `Apex REST Service Response`
    - `HTTP Request` is now `HTTP Callout Request`
    - `HTTP Response` is now `HTTP Callout Response`

    ![image](https://github.com/jongpie/NebulaLogger/assets/1267157/44a3b992-6ce9-4bd6-847f-d5e8cbb1fea3)

## Reduced the Max Length of Several Existing Long Textarea Fields

As Nebula Logger's data model continues to grow, the per-object limit for long textarea fields becomes increasingly important to consider. To compensate for this limit, and to better reflect the amount of data typically stored in these fields, several older fields have been updated to reduce their max length:
- `LogEntryEvent__e.ExceptionMessage__c`: 131,072 limit reduced to 5,000
- `LogEntryEvent__e.ExceptionStackTrace__c`: 131,072 limit reduced to 5,000
- `LogEntryEvent__e.StackTrace__c`: 131,072 limit reduced to 5,000
- `LogEntryEvent__e.Topics__c` (deprecated/no longer used): 131,072 limit reduced to 256
- `LogEntry__c.ExceptionMessage__c`: 131,072 limit reduced to 5,000
- `LogEntry__c.ExceptionStackTrace__c`: 131,072 limit reduced to 5,000
- `LogEntry__c.FlowDescription__c`: 131,072 limit reduced to 5,000
- `LogEntry__c.StackTrace__c`: 131,072 limit reduced to 5,000

# Installation Info

## Core Unlocked Package - no namespace

Full Changelog: https://github.com/jongpie/NebulaLogger/compare/v4.12.8...v4.12.9

- SF CLI: `sf package install --wait 20 --security-type AdminsOnly --package 04t5Y000001Mk7BQAS`
- SFDX CLI: `sfdx force:package:install --wait 20 --securitytype AdminsOnly --package 04t5Y000001Mk7BQAS`
- Sandbox: https://test.salesforce.com/packaging/installPackage.apexp?p0=04t5Y000001Mk7BQAS
- Production: https://login.salesforce.com/packaging/installPackage.apexp?p0=04t5Y000001Mk7BQAS

<hr /><em>This discussion was created from the release <a href='https://github.com/jongpie/NebulaLogger/releases/tag/v4.12.9'>Added RestRequest and RestResponse Logging</a>.</em>

**Comment by aashams1212** [comment]
@jongpie How can we use the new methods setRestRequestDetails. I can't see any traces of it in logger?

**Reply by jongpie** [reply]
Hi @aashams1212 - great question, I'll have to add some better examples to the docs for this functionality. In the meantime, you can add something like one of these examples to your code to log the REST request & REST response details:

```apex
// One option: Log your own instances of a request & response
System.RestRequest request; // TODO provide your own instance
System.RestResponse response;  // TODO provide your own instance
Logger.info('Here is 1 log entry, with details for RestRequest & RestResponse')
      .setRestRequestDetails(request)
      .setRestResponseDetails(response);
Logger.saveLog();

// Another option: Log Apex's static instances of request & response
Logger.info('Here is 1 log entry, with details for RestRequest & RestResponse')
      .setRestRequestDetails(System.RestContext.request)
      .setRestResponseDetails(System.RestContext.response);
Logger.saveLog();
```

Hope this helps! Let me know if you have any other questions.
